lang/java/idl/src/main/java/org/apache/avro/idl/IdlFile.java

```diff
@@ -0,0 +1,120 @@
+ * limitations under the License.
+ */
+package org.apache.avro.idl;
+
+import org.apache.avro.Protocol;
+import org.apache.avro.Schema;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * A parsed IdlFile. Provides access to the named schemas in the IDL file and
+ * the protocol containing the schemas.
+ */
+public class IdlFile {
+  private final Protocol protocol;
+  private final String namespace;
+  private final Map<String, Schema> namedSchemas;
+  private final List<String> warnings;
+
+  IdlFile(Protocol protocol, List<String> warnings) {
+    this(protocol.getNamespace(), protocol.getTypes(), protocol, warnings);
+  }
+
+  private IdlFile(String namespace, Iterable<Schema> schemas, Protocol protocol, List<String> warnings) {
+    this.namespace = namespace;
+    this.namedSchemas = new LinkedHashMap<>();
+    for (Schema namedSchema : schemas) {
+      this.namedSchemas.put(namedSchema.getFullName(), namedSchema);
+    }
+    this.protocol = protocol;
+    this.warnings = Collections.unmodifiableList(new ArrayList<>(warnings));
+  }
+
+  /**
+   * The protocol defined by the IDL file.
+   */
+  public Protocol getProtocol() {
+    return protocol;
+  }
+
+  public List<String> getWarnings() {
+    return warnings;
+  }
+
+  public List<String> getWarnings(String importFile) {
+    return warnings.stream()
+        .map(warning -> importFile + " " + Character.toLowerCase(warning.charAt(0)) + warning.substring(1))
+        .collect(Collectors.toList());
+  }
+
+  /**
+   * The default namespace to resolve schema names against.
+   */
+  public String getNamespace() {
+    return namespace;
+  }
+
+  /**
+   * The named schemas defined by the IDL file, mapped by their full name.
+   */
+  public Map<String, Schema> getNamedSchemas() {
+    return Collections.unmodifiableMap(namedSchemas);
+  }
+
+  /**
+   * Get a named schema defined by the IDL file, by name. The name can be a simple
+   * name in the default namespace of the IDL file (e.g., the namespace of the
+   * protocol), or a full name.
+   *
+   * @param name the full name of the schema, or a simple name
+   * @return the schema, or {@code null} if it does not exist
+   */
+  public Schema getNamedSchema(String name) {
+    Schema result = namedSchemas.get(name);
+    if (result != null) {
+      return result;
+    }
+    if (namespace != null && !name.contains(".")) {
+      result = namedSchemas.get(namespace + "." + name);
+    }
+    return result;
+  }
+
+  // Visible for testing
+  String outputString() {
+    if (protocol != null) {
+      return protocol.toString();
+    }
+    if (namedSchemas.isEmpty()) {
+      return "[]";
+    } else {
+      StringBuilder buffer = new StringBuilder();
+      for (Schema schema : namedSchemas.values()) {
+        buffer.append(",").append(schema);
```

Comment:
```suggestion
        buffer.append(',').append(schema);
```
nit: it is a bit friendlier to memory allocation